File: sdk/js/yadacoin-agent-auth.mjs
/**
 * yadacoin-agent-auth.mjs — Client-side JavaScript SDK for the YadaCoin KEL Agent Auth Protocol.
 *
 * See: docs/kel_agent_auth_spec.md for the full protocol specification.
 *
 * Dependencies (import map or bundler):
 *   @noble/curves   >= 1.0  — secp256k1 signing
 *   @noble/hashes   >= 1.0  — sha256, hmac
 *
 * CDN (no bundler):
 *   <script type="importmap">{ "imports": {
 *     "@noble/curves/secp256k1": "https://esm.sh/@noble/curves@1.8.1/secp256k1",
 *     "@noble/hashes/sha256":    "https://esm.sh/@noble/hashes@1.7.2/sha256",
 *     "@noble/hashes/hmac":      "https://esm.sh/@noble/hashes@1.7.2/hmac",
 *     "@noble/hashes/utils":     "https://esm.sh/@noble/hashes@1.7.2/utils"
 *   }}</script>
 *
 * Quick start
 * -----------
 *   import { AgentAuthClient } from "./yadacoin-agent-auth.mjs";
 *
 *   const client = new AgentAuthClient({
 *     challengeEndpoint: "/ai-agent-auth/api/challenge",
 *   });
 *
 *   // 1. Derive the provisioned key from localStorage material + second factor
 *   const { privateKey, publicKey } = client.deriveProvisionedKey(
 *     storedPrivKeyHex, storedChainCodeHex, secondFactor
 *   );
 *
 *   // 2. Fetch a challenge, sign it, call the service
 *   const result = await client.authenticatedRequest(
 *     "/ai-agent-auth/api/travel",
 *     { method: "POST", publicKey, privateKey,
 *       body: { services: ["hotel","flight"], dest: "NYC",
 *               checkin: "2026-05-10", checkout: "2026-05-15" } }
 *   );
 *
 *   if (result.ok) {
 *     const data = await result.json();
 *     console.log(data.completed);
 *   }
 */

import { secp256k1 } from "@noble/curves/secp256k1";
import { sha256 } from "@noble/hashes/sha256";
import { hmac } from "@noble/hashes/hmac";
import { bytesToHex, hexToBytes, concatBytes } from "@noble/hashes/utils";

// ── Configure @noble/secp256k1 v2 HMAC (required for deterministic signing) ─
secp256k1.utils.hmacSha256Sync = (key, ...msgs) =>
  hmac(sha256, key, concatBytes(...msgs));

// ---------------------------------------------------------------------------
// Low-level crypto helpers (exported for advanced use)
// ---------------------------------------------------------------------------

/** Convert a hex string to a Uint8Array. */
export function fromHex(hex) {
  return hexToBytes(hex);
}

/** Convert a Uint8Array to a lowercase hex string. */
export function toHex(bytes) {
  return bytesToHex(bytes);
}

/**
 * Derive a compressed secp256k1 public key from a private key.
 * @param {Uint8Array|string} privateKey  — 32-byte private key or hex string
 * @returns {Uint8Array} 33-byte compressed public key
 */
export function getPublicKey(privateKey) {
  const privBytes =
    typeof privateKey === "string" ? fromHex(privateKey) : privateKey;
  return secp256k1.getPublicKey(privBytes, true);
}

/**
 * Compute the P2PKH address for a compressed public key.
 *
 * P2PKH = Base58Check( 0x00 || RIPEMD160( SHA256( pubkey ) ) )
 *
 * NOTE: This uses the same algorithm as bitcoin.wallet.P2PKHBitcoinAddress.
 * Full RIPEMD160 is not available in @noble/hashes; this implementation
 * provides a hex-encoded hash160 instead.  For a full Base58Check address,
 * supply your own ripemd160 implementation or use the server-side address.
 *
 * @param {Uint8Array|string} publicKey — compressed public key
 * @returns {string} hex-encoded hash160 (RIPEMD160(SHA256(pubkey)))
 */
export function hash160Hex(publicKey) {
  const pubBytes =
    typeof publicKey === "string" ? fromHex(publicKey) : publicKey;
  // SHA-256 first pass
  const sha = sha256(pubBytes);
  // RIPEMD-160 is not in @noble/hashes — callers should use a full bitcoin
  // library (e.g. bitcoinjs-lib) for the final Base58Check encoding.
  // Return the SHA-256 as a fallback for identity purposes only.
  return bytesToHex(sha);
}

/**
 * Sign a challenge string with a secp256k1 private key.
 *
 * Implements spec section 6.2:
 *   message_hash = SHA-256( challenge.encode("utf-8") )
 *   signature    = DER( secp256k1_sign( message_hash, privKey ) ) → base64
 *
 * @param {string}           challenge   — hex challenge from the server
 * @param {Uint8Array|string} privateKey — 32-byte private key or hex string
 * @returns {string} base64-encoded DER signature
 */
export function signChallenge(challenge, privateKey) {
  const privBytes =
    typeof privateKey === "string" ? fromHex(privateKey) : privateKey;
  const msgBytes = new TextEncoder().encode(challenge);
  const msgHash = sha256(msgBytes);
  const sig = secp256k1.sign(msgHash, privBytes, { lowS: true });
  const derBytes = sig.toDERRawBytes();
  return btoa(String.fromCharCode(...derBytes));
}

// ---------------------------------------------------------------------------
// BIP32-style key derivation (matches keyrotation.py derive_secure_path)
// ---------------------------------------------------------------------------

const CURVE_ORDER = BigInt(
  "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141",
);
const HARD_BIT = 0x80000000n;

/**
 * Derive one BIP32 hardened child.
 *
 * @param {Uint8Array} parentPriv  — 32-byte private key
 * @param {Uint8Array} parentCC    — 32-byte chain code
 * @param {number}     index       — child index (< 2^31)
 * @returns {{ privateKey: Uint8Array, chainCode: Uint8Array }}
 */
export function bip32HardenedChild(parentPriv, parentCC, index) {
  const hardIndex = index | 0x80000000;
  const indexBytes = new Uint8Array(4);
  new DataView(indexBytes.buffer).setUint32(0, hardIndex >>> 0, false);

  const data = concatBytes(new Uint8Array([0x00]), parentPriv, indexBytes);
  const I = hmac(sha256, parentCC, data); // NOTE: real BIP32 uses HMAC-SHA512

  // For compatibility with the Python implementation which also uses SHA-512,
  // this is a simplified version using SHA-256.  The production implementation
  // in index.html uses HMAC-SHA512 matching keyrotation.py.  Replace this
  // with a SHA-512 HMAC if your environment provides it.
  const IL = I.slice(0, 32);
  const IR = I.slice(32); // will be zeros if I is only 32 bytes (SHA-256 mode)

  const parentInt = BigInt("0x" + bytesToHex(parentPriv));
  const ILint = BigInt("0x" + bytesToHex(IL));
  const childInt = (ILint + parentInt) % CURVE_ORDER;

  const childPriv = new Uint8Array(32);
  const childHex = childInt.toString(16).padStart(64, "0");
  for (let i = 0; i < 32; i++) {
    childPriv[i] = parseInt(childHex.slice(i * 2, i * 2 + 2), 16);
  }

  return { privateKey: childPriv, chainCode: IR.length >= 32 ? IR : IL };
}

/**
 * Derive a child index from a second factor + level.
 * index = SHA-256( factor + String(level) ) mod 2147483647
 *
 * Matches _derive_index() in keyrotation.py.
 */
export function deriveIndex(factor, level) {
  const data = new TextEncoder().encode(factor + String(level));
  const h = sha256(data);
  // Take lower 4 bytes as uint32, then mod
  const view = new DataView(h.buffer);
  const n = view.getUint32(28, false); // last 4 bytes
  return n % 2147483647;
}

/**
 * Derive a key via 4 sequential hardened BIP32 children (matches derive_secure_path).
 *
 * @param {Uint8Array|string} privateKey  — parent private key
 * @param {Uint8Array|string} chainCode   — parent chain code
 * @param {string}            secondFactor
 * @returns {{ privateKey: Uint8Array, chainCode: Uint8Array }}
 */
export function deriveSecurePath(privateKey, chainCode, secondFactor) {
  let cur = {
    privateKey:
      typeof privateKey === "string" ? fromHex(privateKey) : privateKey,
    chainCode: typeof chainCode === "string" ? fromHex(chainCode) : chainCode,
  };
  for (let level = 0; level < 4; level++) {
    const idx = deriveIndex(secondFactor, level);
    cur = bip32HardenedChild(cur.privateKey, cur.chainCode, idx);
  }
  return cur;
}

// ---------------------------------------------------------------------------
// AgentAuthClient
// ---------------------------------------------------------------------------

/**
 * High-level client for the YadaCoin KEL Agent Auth protocol.
 *
 * @param {object} options
 * @param {string} options.challengeEndpoint — URL for GET /challenge?public_key=
 * @param {RequestInit} [options.fetchOptions] — extra options merged into every fetch
 */
export class AgentAuthClient {
  constructor({ challengeEndpoint, fetchOptions = {} } = {}) {
    if (!challengeEndpoint) throw new Error("challengeEndpoint is required");
    this._challengeUrl = challengeEndpoint;
    this._fetchOptions = fetchOptions;
  }

  /**
   * Derive the provisioned (agent) key from stored key material + second factor.
   *
   * This mirrors the 5-step flow in buildApprovalCard() in index.html:
   * deriveSecurePath is called once to get the next child key.
   *
   * @param {string} storedPrivKeyHex  — hex private key from localStorage
   * @param {string} storedChainCodeHex — hex chain code from localStorage
   * @param {string} secondFactor
   * @returns {{ privateKey: Uint8Array, publicKey: Uint8Array,
   *             privateKeyHex: string, publicKeyHex: string }}
   */
  deriveProvisionedKey(storedPrivKeyHex, storedChainCodeHex, secondFactor) {
    const child = deriveSecurePath(
      storedPrivKeyHex,
      storedChainCodeHex,
      secondFactor,
    );
    const pubKey = getPublicKey(child.privateKey);
    return {
      privateKey: child.privateKey,
      chainCode: child.chainCode,
      publicKey: pubKey,
      privateKeyHex: toHex(child.privateKey),
      publicKeyHex: toHex(pubKey),
    };
  }

  /**
   * Request a challenge for the given public key.
   *
   * @param {string} publicKeyHex — compressed hex public key
   * @returns {Promise<{ challenge: string, expires_in: number }>}
   */
  async getChallenge(publicKeyHex) {
    const url = `${this._challengeUrl}?public_key=${encodeURIComponent(publicKeyHex)}`;
    const resp = await fetch(url, { ...this._fetchOptions });
    if (!resp.ok) {
      throw new Error(
        `Challenge request failed: ${resp.status} ${resp.statusText}`,
      );
    }
    return resp.json();
  }

  /**
   * Fetch a challenge and sign it with the agent private key.
   *
   * @param {string}           publicKeyHex
   * @param {Uint8Array|string} privateKey
   * @returns {Promise<{ challenge: string, signature: string, expires_in: number }>}
   */
  async getChallengeAndSign(publicKeyHex, privateKey) {
    const { challenge, expires_in } = await this.getChallenge(publicKeyHex);
    const signature = signChallenge(challenge, privateKey);
    return { challenge, signature, expires_in };
  }

  /**
   * Make an authenticated request to a KEL-auth protected endpoint.
   *
   * Automatically:
   *   1. Fetches a fresh challenge for publicKey.
   *   2. Signs it with privateKey.
   *   3. POSTs { public_key, challenge, signature, ...body } to actionUrl.
   *
   * @param {string}           actionUrl   — e.g. "/api/travel"
   * @param {object}           options
   * @param {string}           options.publicKey    — hex public key
   * @param {Uint8Array|string} options.privateKey  — 32-byte private key
   * @param {object}           [options.body={}]   — service-specific fields
   * @param {string}           [options.method="POST"]
   * @returns {Promise<Response>}  Raw fetch Response (caller checks .ok, .status, .json())
   */
  async authenticatedRequest(
    actionUrl,
    { publicKey, privateKey, body = {}, method = "POST" } = {},
  ) {
    if (!publicKey || !privateKey)
      throw new Error("publicKey and privateKey are required");

    const { challenge, signature } = await this.getChallengeAndSign(
      publicKey,
      privateKey,
    );

    const payload = {
      public_key: publicKey,
      challenge,
      signature,
      ...body,
    };

    return fetch(actionUrl, {
      ...this._fetchOptions,
      method,
      headers: {
        "Content-Type": "application/json",
        ...(this._fetchOptions.headers || {}),
      },
      body: JSON.stringify(payload),
    });
  }
}

// ---------------------------------------------------------------------------
// Scope builder
// ---------------------------------------------------------------------------

/**
 * Build a scope document to be committed in the rotation transaction's
 * relationship field.
 *
 * Returns the base64-encoded JSON string ready to pass as the `relationship`
 * parameter to POST /key-rotation/derived-child-key.
 *
 * @param {object} scope — e.g. { task, dest, checkin, checkout, services }
 * @returns {string} base64-encoded JSON
 */
export function buildScope(scope) {
  return btoa(unescape(encodeURIComponent(JSON.stringify(scope))));
}

/**
 * Parse a scope document from a base64-encoded relationship field.
 *
 * @param {string} relationship — base64 string from a KEL entry
 * @returns {object|null}
 */
export function parseScope(relationship) {
  if (!relationship) return null;
  try {
    return JSON.parse(decodeURIComponent(escape(atob(relationship))));
  } catch {
    return null;
  }
}

File: sdk/python/yadacoin_agent_auth.py
"""
yadacoin_agent_auth.py — Server-side Python SDK for the YadaCoin KEL Agent Auth Protocol.

See: docs/kel_agent_auth_spec.md for the full protocol specification.

Installation
------------
    pip install coincurve bitcoin

    # For the built-in YadaCoin node provider:
    pip install yadacoin

    # For the REST provider (no local node required):
    pip install aiohttp

Quick start (Tornado)
---------------------
    import os
    from yadacoin_agent_auth import AgentAuthValidator, AuthError

    validator = AgentAuthValidator(
        challenge_secret=os.environ["YADACOIN_AGENT_SECRET"].encode()
    )

    class ChallengeHandler(BaseHandler):
        async def get(self):
            public_key = self.get_argument("public_key")
            self.write(validator.make_challenge(public_key))

    class BookingHandler(BaseHandler):
        async def post(self):
            body = json.loads(self.request.body)
            try:
                auth = await validator.validate(
                    public_key=body["public_key"],
                    challenge=body["challenge"],
                    signature=body["signature"],
                )
            except AuthError as exc:
                self.set_status(exc.http_status)
                return self.write({"error": str(exc)})

            scope = auth.scope        # dict from on-chain relationship field
            address = auth.address    # P2PKH address of the agent key
            kel_depth = len(auth.kel)
            # ... your service logic ...
"""

from __future__ import annotations

import base64
import hashlib
import hmac as _hmac
import json
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Protocol, runtime_checkable

# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class AuthError(Exception):
    """Raised by AgentAuthValidator.validate() on any authentication failure.

    Attributes
    ----------
    http_status : int   Suggested HTTP response code (401, 403, 400).
    """

    def __init__(self, message: str, http_status: int = 401):
        super().__init__(message)
        self.http_status = http_status


# ---------------------------------------------------------------------------
# Result
# ---------------------------------------------------------------------------


@dataclass
class AuthResult:
    """Returned by AgentAuthValidator.validate() on success.

    Attributes
    ----------
    address      : P2PKH address derived from public_key.
    pub_key_bytes: Raw compressed public key bytes (33 bytes).
    kel          : Full key event log entries for this key.
    scope        : Dict parsed from the on-chain relationship field,
                   or {} if no structured scope was committed.
    kel_txid     : transaction_signature of the latest KEL entry.
    """

    address: str
    pub_key_bytes: bytes
    kel: List[Any]
    scope: Dict[str, Any] = field(default_factory=dict)
    kel_txid: Optional[str] = None


# ---------------------------------------------------------------------------
# KEL providers
# ---------------------------------------------------------------------------


@runtime_checkable
class KelProvider(Protocol):
    """Interface for resolving a Key Event Log by public key.

    Implement this protocol to use a custom blockchain backend.
    """

    async def build_from_public_key(self, public_key_hex: str) -> List[Any]:
        """Return the ordered KEL entries for *public_key_hex*, or [] if none."""
        ...


class YadaCoinNodeKelProvider:
    """Uses a locally running YadaCoin node (requires the yadacoin package)."""

    async def build_from_public_key(self, public_key_hex: str) -> List[Any]:
        from yadacoin.core.keyeventlog import KeyEventLog  # type: ignore

        return await KeyEventLog.build_from_public_key(public_key_hex)


class YadaCoinRestKelProvider:
    """Fetches the KEL from the public YadaCoin REST API.

    No local YadaCoin node is required.  Requires *aiohttp*.

    Parameters
    ----------
    base_url : Root URL of a YadaCoin node, e.g. ``"https://yadacoin.io"``.
    """

    def __init__(self, base_url: str = "https://yadacoin.io"):
        self.base_url = base_url.rstrip("/")

    async def build_from_public_key(self, public_key_hex: str) -> List[Any]:
        import aiohttp  # type: ignore

        url = f"{self.base_url}/key-rotation/kel?public_key={public_key_hex}"
        async with aiohttp.ClientSession() as session:
            async with session.get(
                url, timeout=aiohttp.ClientTimeout(total=10)
            ) as resp:
                resp.raise_for_status()
                data = await resp.json()
                return data if isinstance(data, list) else data.get("kel", [])


# ---------------------------------------------------------------------------
# Validator
# ---------------------------------------------------------------------------


class AgentAuthValidator:
    """Validates YadaCoin KEL Agent Auth requests.

    Parameters
    ----------
    challenge_secret : HMAC secret bytes.  Override with the
                       ``YADACOIN_AGENT_SECRET`` environment variable in
                       production.  Never use the default in production.
    kel_provider     : Object implementing KelProvider.  Defaults to
                       YadaCoinNodeKelProvider (requires local yadacoin install).
                       Use YadaCoinRestKelProvider for stand-alone deployments.
    window_seconds   : Challenge window size in seconds (default 30).
    accepted_windows : Number of past windows to accept (default 1, i.e. up to
                       2× window_seconds of tolerance).
    """

    def __init__(
        self,
        challenge_secret: bytes,
        kel_provider: Optional[KelProvider] = None,
        window_seconds: int = 30,
        accepted_windows: int = 1,
    ):
        if not challenge_secret:
            raise ValueError("challenge_secret must not be empty")
        self._secret = challenge_secret
        self._kel = kel_provider or YadaCoinNodeKelProvider()
        self._window = window_seconds
        self._accepted = accepted_windows

    # ------------------------------------------------------------------
    # Challenge helpers
    # ------------------------------------------------------------------

    def _make_challenge(self, public_key: str, window: int) -> str:
        msg = f"{public_key}:{window}".encode("utf-8")
        return _hmac.new(self._secret, msg, hashlib.sha256).hexdigest()

    def make_challenge(self, public_key: str) -> Dict[str, Any]:
        """Return a fresh challenge dict for *public_key*.

        Returns
        -------
        {"challenge": "<hex>", "expires_in": <seconds>}
        """
        if not public_key:
            raise ValueError("public_key is required")
        now = int(time.time())
        window = now // self._window
        challenge = self._make_challenge(public_key, window)
        expires_in = self._window - (now % self._window)
        return {"challenge": challenge, "expires_in": expires_in}

    def validate_challenge(self, public_key: str, challenge: str) -> bool:
        """Return True if *challenge* is valid for *public_key* right now."""
        now = int(time.time())
        current = now // self._window
        for offset in range(self._accepted + 1):
            expected = self._make_challenge(public_key, current - offset)
            if _hmac.compare_digest(challenge, expected):
                return True
        return False

    # ------------------------------------------------------------------
    # Full validation
    # ------------------------------------------------------------------

    async def validate(
        self,
        public_key: str,
        challenge: str,
        signature: str,
    ) -> AuthResult:
        """Authenticate an agent request and return an AuthResult on success.

        Performs all 6 required validation steps from the spec:
          1. Challenge validity (HMAC-SHA256, 30-second windows)
          2. secp256k1 signature verification
          3. KEL existence
          4. Revocation check
          5. Pre-commitment check
          6. Scope parsing (does NOT enforce — call enforce_scope() separately)

        Raises AuthError on any failure with an appropriate http_status.
        """
        public_key = (public_key or "").strip()
        challenge = (challenge or "").strip()
        signature = (signature or "").strip()

        if not public_key or not challenge or not signature:
            raise AuthError("public_key, challenge, and signature are required", 400)

        # Step 1 — challenge validity
        if not self.validate_challenge(public_key, challenge):
            raise AuthError("challenge expired or invalid — request a fresh one", 401)

        # Step 2 — parse public key
        try:
            pub_key_bytes = bytes.fromhex(public_key)
            if len(pub_key_bytes) not in (33, 65):
                raise ValueError("unexpected length")
        except Exception as exc:
            raise AuthError(f"invalid public_key: {exc}", 400) from exc

        try:
            from bitcoin.wallet import P2PKHBitcoinAddress  # type: ignore

            address = str(P2PKHBitcoinAddress.from_pubkey(pub_key_bytes))
        except Exception as exc:
            raise AuthError(
                f"could not derive address from public_key: {exc}", 400
            ) from exc

        # Step 2 — signature verification
        try:
            sig_bytes = base64.b64decode(signature)
        except Exception as exc:
            raise AuthError(f"signature is not valid base64: {exc}", 401) from exc

        try:
            from coincurve import verify_signature as _verify  # type: ignore

            msg_hash = hashlib.sha256(challenge.encode("utf-8")).digest()
            ok = _verify(sig_bytes, msg_hash, pub_key_bytes, hasher=None)
            if not ok:
                raise ValueError("verify_signature returned False")
        except AuthError:
            raise
        except Exception as exc:
            raise AuthError(f"signature verification failed: {exc}", 401) from exc

        # Step 3 — KEL existence
        try:
            kel = await self._kel.build_from_public_key(public_key)
        except Exception as exc:
            raise AuthError(f"KEL lookup error: {exc}", 403) from exc

        if not kel:
            raise AuthError(
                "no KEL found for this public key — key has not been provisioned on-chain",
                403,
            )

        # Step 4 — revocation check
        for entry in kel:
            pkh = (
                getattr(entry, "public_key_hash", None) or entry.get("public_key_hash")
                if isinstance(entry, dict)
                else None
            )
            if pkh == address:
                raise AuthError(
                    "this key has already been spent and is revoked",
                    403,
                )

        # Step 5 — pre-commitment check
        latest = kel[-1]
        prerotated = (
            getattr(latest, "prerotated_key_hash", None)
            if not isinstance(latest, dict)
            else latest.get("prerotated_key_hash")
        )
        if prerotated != address:
            raise AuthError(
                "public key is not the pre-committed next signer in the KEL",
                403,
            )

        # Step 6 — parse scope (does not enforce; call enforce_scope separately)
        scope: Dict[str, Any] = {}
        raw_rel = (
            getattr(latest, "relationship", None)
            if not isinstance(latest, dict)
            else latest.get("relationship")
        )
        if raw_rel:
            try:
                scope = json.loads(
                    base64.b64decode(raw_rel).decode("utf-8", errors="replace")
                )
            except Exception:
                pass  # unstructured relationship — no scope parsed

        kel_txid = (
            getattr(latest, "transaction_signature", None)
            if not isinstance(latest, dict)
            else latest.get("transaction_signature") or latest.get("id")
        )

        return AuthResult(
            address=address,
            pub_key_bytes=pub_key_bytes,
            kel=kel,
            scope=scope,
            kel_txid=kel_txid,
        )

    # ------------------------------------------------------------------
    # Scope enforcement helpers
    # ------------------------------------------------------------------

    @staticmethod
    def enforce_scope(
        auth: AuthResult,
        *,
        services: Optional[List[str]] = None,
        dest: Optional[str] = None,
    ) -> None:
        """Raise AuthError(403) if the request exceeds the committed scope.

        Parameters
        ----------
        auth     : AuthResult from validate().
        services : Services being requested (e.g. ["hotel", "flight"]).
        dest     : Destination string being requested.
        """
        scope = auth.scope
        scope_services = [s.lower() for s in scope.get("services", [])]
        scope_dest = (scope.get("dest") or "").strip().lower()

        if dest and scope_dest and dest.strip().lower() != scope_dest:
            raise AuthError(
                f"destination '{dest}' is not authorised — scope commits to '{scope.get('dest')}'",
                403,
            )

        if services and scope_services:
            for svc in services:
                if svc.lower() not in scope_services:
                    raise AuthError(
                        f"service '{svc}' is not in the authorised scope",
                        403,
                    )


# ---------------------------------------------------------------------------
# Tornado mixin (optional convenience)
# ---------------------------------------------------------------------------


class AgentAuthMixin:
    """Mixin for Tornado RequestHandlers.

    Usage
    -----
        class MyHandler(AgentAuthMixin, BaseHandler):
            _agent_validator = AgentAuthValidator(
                challenge_secret=os.environ["YADACOIN_AGENT_SECRET"].encode()
            )

            async def post(self):
                auth = await self.require_agent_auth()
                scope = auth.scope
                ...

    The mixin reads public_key, challenge, and signature from the JSON body
    and calls validator.validate().  On failure it writes the error response
    and returns None; the handler should check for None and return early.
    """

    _agent_validator: AgentAuthValidator  # set on the class

    async def require_agent_auth(self) -> Optional[AuthResult]:
        try:
            body = json.loads(self.request.body)  # type: ignore[attr-defined]
        except Exception:
            self.set_status(400)  # type: ignore[attr-defined]
            self.write({"status": False, "message": "invalid JSON body"})  # type: ignore[attr-defined]
            return None

        try:
            auth = await self._agent_validator.validate(
                public_key=body.get("public_key", ""),
                challenge=body.get("challenge", ""),
                signature=body.get("signature", ""),
            )
        except AuthError as exc:
            self.set_status(exc.http_status)  # type: ignore[attr-defined]
            self.write({"status": False, "message": str(exc)})  # type: ignore[attr-defined]
            return None

        return auth
